### ISSUE TITLE
Q: How easy would it be to use this to use NVidia Tensorrt from C#? 

### ISSUE THREAD

**Issue by nietras** [InfectedLibraries/Biohazrd#83]
Hi, this project looks really interesting. 👍 I want to use Tensorrt from c#, but as is unfortunately often the case there is no C API so simple p/invoke is not an option. How easy do you think it would be to wrap this using this project? For Windows.

**Comment by PathogenDavid** [comment]
Thanks for the interest! 💜

Short version: I think it should be pretty easy! Honestly the biggest problem if you were to do this yourself would probably be the lack of documentation.

--------------

I've never used TensorRT before (I'm not much of an machine learning guy), but looking through the API suggests this library would be trivial for Biohazrd to translate.

I have a (yet-to-be-published) tool that automatically assesses how compatible a given library is with Biohazrd, and it had a favorable outlook. ([If you want to look at the report, I've uploaded it here](https://sphere.pathogenstudios.com/pgnwww/temp/Report_TensorRT.html))

To summarize the results:

* Use of `std::vector`
  * [`nvonnxparser::IParser::supportsModel`](https://github.com/NVIDIA/TensorRT/blob/492878b2df3c9bf19a26c2b66cba129f450475bc/include/NvOnnxParser.h#L149-L160) - In theory translating `std::vector` is possible, but it is not something that appears on the public API surface on any of the libraries me or my main sponsor are focused on.
    * You may have noticed `NvOnnxParser.h` was skipped. This is because parsing this file is blocked by an external issue https://github.com/microsoft/STL/issues/1300 -- I haven't had a chance to explore workarounds to this issue since it hasn't seriously affected me yet.
    * Note that this means the entire `NvOnnxParser.h` file could not be processed in Biohazrd as-is. You'd either have to figure out a good solution to that issue or modify it to avoid `intrin.h` getting included. Not the worst thing ever, but I thought I should mention it.
* C++ templates
  * [`EnumMax<T>`](https://github.com/NVIDIA/TensorRT/blob/492878b2df3c9bf19a26c2b66cba129f450475bc/include/NvInferRuntimeCommon.h#L100-L101) - Not of concern - This is a C++ template shenanigan to get the max value of an enum. If this were needed I'd just recommend special-casing its translation.
  * [`PluginRegistrar<T>`](https://github.com/NVIDIA/TensorRT/blob/492878b2df3c9bf19a26c2b66cba129f450475bc/include/NvInferRuntimeCommon.h#L1359-L1375) - Not of concern - This is (another) template shenanigan to automatically register plugins when a library is loaded. If you implemented a plugin from C# you'd need to use a different strategy to register it.
* Global variable/constant -- In general none of these are of great concern. In TensorRT these are all constants, which are tracked by #5. After #75 this is pretty viable, although none of these appear to be critical for using the library.
  * [`NV_TENSORRT_VERSION`](https://github.com/NVIDIA/TensorRT/blob/492878b2df3c9bf19a26c2b66cba129f450475bc/include/NvInferRuntimeCommon.h#L85-L86)
  * [`MAX_DIMS`](https://github.com/NVIDIA/TensorRT/blob/492878b2df3c9bf19a26c2b66cba129f450475bc/include/NvInferRuntimeCommon.h#L197) - [Biohazrd translates constant arrays with a `Length` constant field](https://github.com/InfectedLibraries/InfectedImGui/blob/aaabb2ba2b3bc74d20cdd9ff8ccb02ad1ac6d796/InfectedImGui/%23Generated/SynthesizedDeclarations.cs#L89), so this constant isn't really necessary.
  * [`kFORMAT_COMBINATION_LIMIT`](https://github.com/NVIDIA/TensorRT/blob/492878b2df3c9bf19a26c2b66cba129f450475bc/include/NvInferRuntime.h#L450-L453) -- This one may matter, but it could be hard-coded or special-cased for the time being.

I did not check anything in the [`plugin](https://github.com/NVIDIA/TensorRT/tree/492878b2df3c9bf19a26c2b66cba129f450475bc/plugin) directory since I am under the impression that these APIs are not meant to be used directly. If this assessment is incorrect let me know and I can look over them too. (Although at a glance they look like they'd require me to install the CUDA SDK.)

-----------

Outside of that, this stood out to me:

> It is designed to work in a complementary fashion with training frameworks such as TensorFlow, Caffe, PyTorch, MXNet, etc.

Does TensorRT rely on one of these technologies in order to be useful? Would bindings for one of these need to be made for .NET too or do such bindings already exist? (Assuming there aren't already .NET equivalents to any of those.)

--------

For my own reference (because NVIDIA loves to bury this stuff), here's the TensorRT API reference: https://docs.nvidia.com/deeplearning/tensorrt/api/c_api/index.html

**Comment by PathogenDavid** [comment]
> For Windows.

I will note that currently only Windows x64 is supported. Linux x64 is currently planned/sponsored, but everything else is mostly up in the air. (Linux ARM64 is on the very distant roadmap. Anything else would likely require sponsorship.)

Some libraries generated by Biohazrd only support .NET 5 due to the use of C#9 function pointers, and TensorRT would be one of them.

In theory supporting older .NET versions is possible, but I'm not particularly keen on supporting them without a strong incentive. Biohazrd is extensible enough that you could probably add legacy [`UnmanagedFunctionPointer`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.unmanagedfunctionpointerattribute) support yourself, but it might not be pretty.

**Comment by nietras** [comment]
@PathogenDavid thank you very much for the very detailed and thoughtful reply. My apologies for my tardiness. :) This sounds great. Some answers and comments below.

You can see examples of usage in e.g.:

https://github.com/NVIDIA/TensorRT/blob/master/samples/opensource/sampleOnnxMNIST/sampleOnnxMNIST.cpp

> nvonnxparser::IParser::supportsModel -

This shouldn't be needed I think.

>  the entire NvOnnxParser.h file could not be processed in Biohazrd as-is

Good to mention this. Since I need to use this given my models are in ONNX format, it is unfortunately a requirement. 

> PluginRegistrar<T> - Not of concern
> If you implemented a plugin from C# you'd need to use a different strategy to register it.

Don't need to do that. But need to load/use the `nvinfer_plugin.dll` plugin, not sure this is impacted by this issue.

> Global variable/constant

Shouldn't be a problem.

> It is designed to work in a complementary fashion with training frameworks such as TensorFlow, Caffe, PyTorch, MXNet, etc.
> Does TensorRT rely on one of these technologies in order to be useful? Would bindings for one of these need to be made for .NET too or do such bindings already exist? (Assuming there aren't already .NET equivalents to any of those.)

No not at all, TensorRT is useful in it self. I prefer using it via https://github.com/microsoft/onnxruntime but this does not support all the use cases via that. TensorRT is mainly for running "inference" of models, that is after training. It usually provides big performance improvements compared to running inference based using standard CUDA.

> currently only Windows x64 is supported

TensorRT only supports x64 anyway 😅 So no problem.

> Would bindings for one of these need to be made for .NET too or do such bindings already exist?

Bindings to more ML libraries and frameworks for .NET would be great overall! This is lacking a lot for .NET. For some reason a lot of these libraries have bindings to python and this without a C API and apparently this is easier to do for some reason. I don't know why that is the case? If there was a C API everything would be easier, but a lot of ML libraries don't have that. Luckily, ONNX Runtime does, though.

> Some libraries generated by Biohazrd only support .NET 5 due to the use of C#9 function pointers, and TensorRT would be one of them.

Yes, that I expected, but function pointers are more a C# 9 thing that a specific .NET 5 thing I think. If you wrap these with normal calls this should be .NET Standard 2.0 compatible I think. But this can be addressed after.

Do you think it is possible to progress based on this? Could I do or help with something here to try to go forward with this and explore this? :)

**Comment by nietras** [comment]
@PathogenDavid I found it interesting the code contains `SWIG` stuff, so perhaps one can wrap this using SWIG and perhaps that is how python interops with the library...
```cpp
#ifdef SWIG
inline IParser* createParser(nvinfer1::INetworkDefinition* network,
                             nvinfer1::ILogger* logger)
{
    return static_cast<IParser*>(
        createNvOnnxParser_INTERNAL(network, logger, NV_ONNX_PARSER_VERSION));
}
#endif // SWIG
```
https://github.com/NVIDIA/TensorRT/blob/492878b2df3c9bf19a26c2b66cba129f450475bc/include/NvOnnxParser.h#L237-L244

**Comment by nietras** [comment]
It's only defined once so not sure that is true... https://github.com/NVIDIA/TensorRT/search?q=SWIG

**Comment by nietras** [comment]
> SWIG does support C#, but personally I wasn't too impressed with the codegen or the strategies it uses. If I thought SWIG was a good binding generator I probably wouldn't have been working on Biohazrd for so long 

I agree on that. We use CNTK via SWIG generated bindings and there are not exactly optimal in all cases. It works, though. So had thought about using SWIG for Tensorrt too. But SWIG depends on generating a C API in a new library/dll which is annoying, hence why I thought of Biohazrd. 👍

> Downlevel .NET is not a priority for me or my sponsor, but if someone does try to use a Biohazrd-generated library in downlevel .NET I'd definitely be interested to hear how it went.

No problem, we will see 👍 

> aren't in a rush, I could probably find some time soon to get a basic generator going for TensorRT. (I don't mind doing it since the library has such a simple API, but I'm definitely interested in hearing your feedback.)

I'm not in a rush so this sounds great! And I also thought with this simple API it might be good example/use case. Please do not hesitate to contact me if there is anything at all. Thanks!

**Comment by PathogenDavid** [comment]
> I'm not in a rush so this sounds great!

Turns out this plan might've been for the best. I ended up finding a few missing features in Biohazrd that were pretty easy for me to solve, but I would've felt guilty if you had run into them. So in the end it turned out to be a good exercise of Biohazrd regardless.

I've gotten the generator done, and I've ported [the MNIST sample](https://github.com/NVIDIA/TensorRT/tree/f1d38b5456e15e2683a4513ea3cb14ed6d94cf53/samples/opensource/sampleMNIST) as well as the necessary portions of the [common sample code](https://github.com/NVIDIA/TensorRT/tree/f1d38b5456e15e2683a4513ea3cb14ed6d94cf53/samples/common).

The only real unexpected roadblock was the fact that Biohazrd doesn't support namespaces yet. (Things can be in namespaces, but it translates everything into a global namespace.) This hasn't been an issue with PhysX or Dear ImGui, but TensorRT has conflicts between its own namespaces. (For example `nvinfer1::IPluginFactory` and `nvcaffeeparser1::IPluginFactory`.) In theory it was still usable because Biohazrd auto-renamed them to `IPluginFactory_0` and `IPluginFactory_1`, but that's obviously not very usable. I hacked together good enough namespace support specifically for TensorRT, but a proper solution (which was already on my mental roadmap) will still be necessary. (The API looks OK, but the codegen and the way it works at generation time is pretty gross.)

I'm currently [investigating a crash in the sample](https://github.com/InfectedLibraries/InfectedTensorRT/blob/93ddf782426f86fcbddda24bd5dc37afabfb1ec0/InfectedTensorRT.Sample.MNIST/SampleMNIST.cs#L175) that is caused because `IConstantLayer` has no input tensors for some reason. ([Native equivalent location](https://github.com/NVIDIA/TensorRT/blob/f1d38b5456e15e2683a4513ea3cb14ed6d94cf53/samples/opensource/sampleMNIST/sampleMNIST.cpp#L270)) I wanted to run the NVIDIA C++ sample in the debugger and compare values, but for whatever reason the debugger wasn't working with the sample and I didn't have time to investigate yet.

I've given you access to the [WIP repository](https://github.com/InfectedLibraries/InfectedTensorRT/tree/wip) (You have to accept access, check your email.) If you have some time I'd definitely appreciate it if you could look over the ported sample code and see if something sticks out. I'm pretty sure there's just a typo somewhere that I'm not noticing, so another set of eyes would probably help. (Unfortunately this part of the code somehow ended up with almost all of the `BIOQUIRK` notes, so it's a little rough.) Plus I'm definitely out of my element here, I'm not even sure what would indicate how many inputs or outputs the `IConstantLayer` should have. There's some rough instructions in the readme on setting things up to run the sample.

Off the top of my head, here's the main tasks left:

* [ ] Fix the sample
* [ ] Implement a workaround for https://github.com/microsoft/STL/issues/1300 so that we can use `NvOnnxParser.h`
* [ ] Implement one of the samples that uses the Onnx parser.
* [ ] (Maybe) Implement https://github.com/InfectedLibraries/Biohazrd/issues/30 since some of the API is annoying to use without it.
* [ ] Cleaning up the readme and clarifying the license of the sample code.

The MNIST sample is almost entirely a mechanical translation of the C++ sample, so minimal effort went into C#ifying it. I put a bit more effort into C#ifying the common sample code, but for the most part it's the same.

**Comment by nietras** [comment]
> I've gotten the generator done, and I've ported the MNIST sample as well as the necessary portions of the common sample code.

Awesome! :) Since the example isn't the `onnx` sample I referred to, things are a bit different that what I am used. `onnx` file contains both network and weights in the same file making it easier.

> caused because IConstantLayer has no input tensors for some reason.

Haven't seen this before, and just to be clear I don't have a lot of experience using the raw Tensorrt API myself, sorry. Mainly used in via ONNX Runtime or the `trtexec` tool, which can also be used to verify things work on a certain file for example. I could think this might relate to how weights and the mean stuff and network are mapped together. The onnx example doesn't have this rather esoteric bit in it, so think it would be easier to get that working once `NvOnnxParser` can be used.

Overall, what I find it great is that it appears all the config options are available which is the primary driver for my interest in this. In ONNX Runtime one can't define this on your own, it uses a single optimization parameter instead but that means certain things can't be configured. The `trtexec` allows one to play around with that.

I haven't had time to look into this in detail, unfortunately. I, of course, can see, as you already noted, that this is very raw with no C#'ification around life time management etc. But it looks promising to me. `NvOnnxParser` is missing of course, I think that would be the no. 1 priority for me and I believe if we get this the example will be easier to get working given the sample code is simpler for that :)

**Comment by PathogenDavid** [comment]
> Awesome! :) Since the example isn't the `onnx` sample I referred to, things are a bit different that what I am used. `onnx` file contains both network and weights in the same file making it easier.

Yeah, I wanted to start with this sample since NVIDIA seems to present it as the Hello World of the library. That sample will come next.

> Haven't seen this before, and just to be clear I don't have a lot of experience using the raw Tensorrt API myself, sorry.

No worries. I have basically zero practical experience with machine learning, so I have no idea how portable the terminology is.

>  I haven't had time to look into this in detail, unfortunately.

No problem, I just wanted to give you an update. If you find some time to help debug the sample I'd of course appreciate it. (I know you only care about the ONNX bits, but want to be sure the sample is broken because of an error I made while re-writing it and not because of an error within Biohazrd.)

>  I, of course, can see, as you already noted, that this is very raw with no C#'ification around life time management etc.

Yup. Also I feel like I didn't make it super clear earlier: The are plans for how Biohazrd will be able to help* you wrap libraries in a safer manner, but it currently doesn't provide much support out of the box. (Right now the focus is on the unsafe interop layer.) You can of course implement the safe layer yourself (or even use Biohazrd to generate it in your own style), but I'm still shaping how Biohazrd "recommends" (and helps) you do this.

(*Part of the reason this I use the word "help" here is I don't believe there's a one size fits all approach for wrapping native interop. For example, an approach that works well for TensorRT wouldn't work well for PhysX.)

**Comment by PathogenDavid** [comment]
@nietras

I developed a workaround for the `intrin0.h` issue (https://github.com/InfectedLibraries/Biohazrd/issues/98) and ported the ONNX MNIST sample and it works!

![image](https://user-images.githubusercontent.com/278957/100247353-df21bb00-2eff-11eb-99c5-3e6d5e1c857c.png)

Changes are live on the private InfectedTensorRT repo if you want to check them out. I haven't had time to look at the non-ONNX sample again to figure out why it's broken.
